README
This library provides vernacular files containing tactics for
simplifying the proofs of inequalities on expressions of real numbers
for the Coq proof assistant.

This package is free software; you can redistribute it and/or modify it
under the terms of CeCILL-C Free Software License (see the COPYING file).
Main author is Guillaume Melquiond <guillaume.melquiond@inria.fr>.


I. Invocation
-------------

In order to use the tactics of the library, one has to import the
"Interval_tactic" file into a Coq proof script. The main tactic is named
"interval".

The tactic can be applied on a goal of the form "(c1 <= e <= c2)%R" with
"e" an expression involving real-valued operators. Sub-expressions that
are not recognized by the tactic should be either terms "t" appearing in
hypothesis inequalities "(c3 <= t <= c4)%R" or simple integers. The
bounds "c1" "c2" ... are expressions that contain only constant leaves,
e.g. (5 / sqrt (1 + PI))%R.

The complete list of recognized goals is as follows:

  (c1 <= e <= c2)%R
  (e <= c2)%R
  (c1 <= e)%R
  (0 < e)%R
  (e <> 0)%R
  (Rabs e <= c2)%R     handled as (-c2 <= e <= c2)%R
  (e1 <= e2)%R         handled as (e1 - e2 <= 0)%R
  (e1 < e2)%R          handled as (0 < e2 - e1)%R
  (e1 <> e2)%R         handled as (e1 - e2 <> 0)%R

Operators recognized by the tactic are: PI, Ropp, Rabs, Rinv, Rsqr, sqrt,
cos, sin, tan, atan, exp, ln, pow, powerRZ, Rplus, Rminus, Rmult, Rdiv.
There are some restrictions on the domain of a few functions: pow and
powerRZ should be written with a numeric exponent; the input of cos and
sin should be between -2*PI and 2*PI; the input of tan should be between
-PI/2 and PI/2.

The tactic also recognizes integral expressions RInt whose bounds are
constants and whose integrand is an expression containing only constant
leaves except for the integration variable. Some improper integral
expressions RInt_gen are also supported with bounds (at_right 0)
(at_point _) or (at_point _) (Rbar_locally p_infty). The integrand
should be of the form (fun t => f t * g t) with g one of the following
expressions:

  exp (- (_ * t))
  powerRZ t _ * (ln t) ^ _
  / (t * (ln t) ^ _)

A helper tactic "interval_intro e" is also available. Instead of proving
the current goal, it computes an enclosure of the expression "e" passed
as argument and it introduces the inequalities into the proof context.
If only one bound is needed, the keywords "lower" and "upper" can be
passed to the tactic, so that it does not perform useless computations.
For example, "interval_intro e lower" introduces only a floating-point
lower bound of "e" in the context. Unless one uses "as" followed by an
intro pattern, the "interval_intro" tactic generates a fresh name
for the hypothesis added to the context.


II. Fine-tuning
---------------

The behavior of the tactics can be tuned by adding an optional set of
parameters "with (param1, param2, ...)" at the end of the tactics. These
parameters are parsed from left to right: If some parameters are
conflicting, the earlier ones are discarded. Available parameter classes
are: (with the type of their arguments, if any)

  i_prec (p:nat)      sets the precision of the floating-point computations
  i_depth (n:nat)     sets the bisection depth (2^n sub-intervals at most)
  i_bisect (x:R)      splits input interval on x and repeat until proven
  i_bisect_diff (x:R) same as i_bisect, but studies variations along x too
  i_bisect_taylor (x:R) (d:nat)
                      same as i_bisect_diff, but computes degree-d Taylor
                      models instead of performing automatic differentiation
  i_integral_prec (p:nat)
                      sets the target relative accuracy of integral
                      expressions to approximately p bits
  i_integral_width (p:Z)
                      sets the target accuracy of integral expressions
                      to an interval width of 2^p
  i_integral_depth (n:nat)
                      sets the bisection depth for bounding integral
                      expressions (2^n sub-intervals at most)
  i_integral_deg (d:nat)
                      sets the degree of Taylor models for approximating
                      the integrand when bounding integral expressions
  i_native_compute    uses native_compute instead of vm_compute
  i_delay             delays proof checking till Qed, especially useful when
                      experimenting with interval_intro

For both tactics, performing a bisection of depth 1 is not much slower
than performing no bisection. If the current goal can be proven by
"interval" with a bisection of depth n, then increasing the depth to n + 1
will not have any noticeable effect. For "interval_intro", increasing the
depth from n to n + 1 can, however, double the computation time.

Performing an "i_bisect_diff" bisection has a much higher cost per
sub-interval, but it can considerably reduce the amount of sub-intervals
considered. As a consequence, unless there is a huge amount of trivial
propositions to prove, one should use this improved bisection.

If the proof process is still too slow, the "i_bisect_taylor" bisection
can be tried instead, as it usually reduces the number of sub-intervals
much further. In some corner cases though, it will not be able to prove
properties for which "i_bisect_diff" would have succeeded.

By default, the precision of the floating-point computations is 30 bits.
If the user enables a bisection, the default depth is 15 for "interval"
and 5 for "interval_intro". When bounding integral expressions, the
tactics target 10 bits of accuracy by splitting the domain into 2^3
subdomains at most and by using degree-10 Taylor models.


III. Examples
-------------

(** BEGIN **)
Require Import Reals.
Require Import Interval.Interval_tactic.

Open Scope R_scope.

Goal
  forall x, -1 <= x <= 1 ->
  sqrt (1 - x) <= 3/2.
Proof.
  intros.
  interval.
Qed.

Goal
  forall x, -1 <= x <= 1 ->
  sqrt (1 - x) <= 141422/100000.
Proof.
  intros.
  interval.
Qed.

Goal
  forall x, -1 <= x <= 1 ->
  sqrt (1 - x) <= 141422/100000.
Proof.
  intros.
  interval_intro (sqrt (1 - x)) upper as H'.
  apply Rle_trans with (1 := H').
  interval.
Qed.

Goal
  forall x, 3/2 <= x <= 2 ->
  forall y, 1 <= y <= 33/32 ->
  Rabs (sqrt(1 + x/sqrt(x+y)) - 144/1000*x - 118/100) <= 71/32768.
Proof.
  intros.
  interval with (i_prec 19, i_bisect x).
Qed.

Goal
  forall x, 1/2 <= x <= 2 ->
  Rabs (sqrt x - (((((122 / 7397 * x + (-1733) / 13547) * x
                   + 529 / 1274) * x + (-767) / 999) * x
                   + 407 / 334) * x + 227 / 925))
    <= 5/65536.
Proof.
  intros.
  interval with (i_bisect_taylor x 3).
Qed.

Goal
  forall x, -1 <= x ->
  x < 1 + powerRZ x 3.
Proof.
  intros.
  interval with (i_bisect_diff x).
Qed.

Require Import Coquelicot.Coquelicot.

Goal
  Rabs (RInt (fun x => atan (sqrt (x*x + 2)) / (sqrt (x*x + 2) * (x*x + 1))) 0 1
        - 5/96*PI*PI) <= 1/1000.
Proof.
  interval with (i_integral_prec 9, i_integral_depth 1, i_integral_deg 5).
Qed.

Goal
  RInt_gen (fun x => 1 * (powerRZ x 3 * ln x^2))
           (at_right 0) (at_point 1) = 1/32.
Proof.
  refine ((fun H => Rle_antisym _ _ (proj2 H) (proj1 H)) _).
  interval.
Qed.

Goal
  Rabs (RInt_gen (fun t => 1/sqrt t * exp (-(1*t)))
                 (at_point 1) (Rbar_locally p_infty)
        - 2788/10000) <= 1/1000.
Proof.
  interval.
Qed.
(*** END ***)


NEWS
Version 3.3.0:
 - added option i_integral_width for absolute width of integrals
 - improved accuracy for interval cos and sin away from zero
 - ensured compatibility from Coq 8.5 to Coq 8.7

Version 3.2.0:
 - added support for some improper integrals using RInt_gen

Version 3.1.1:
 - ensured compatibility with Coq 8.6

Version 3.1.0:
 - improved tactic so that it can be used with backtracking tacticals (try, ||)
 - fixed ineffective computation of integrals with reversed or overlapping bounds

Version 3.0.0:
 - added support for integrals using Coquelicot's RInt operator
 - improved support for Taylor models

Version 2.2.1:
 - moved to MathComp 1.6

Version 2.2.0:
 - improved tactic so that it handles goals with non floating-point bounds
 - added a dependency on Coquelicot to remove an assumption about ln

Version 2.1.0:
 - moved to Flocq 2.5

Version 2.0.0:
 - added support for Taylor models (i_bisect_taylor)
 - added support for logarithm
 - improved tactic so that it handles hypotheses with non floating-point bounds

Version 1.1.0:
 - moved to Flocq 2.4
 - added support for disequalities to the tactic
 - added support for the PI symbol
 - enlarged the domain of the interval versions of cos and sin

Version 1.0.0:
 - removed remaining axioms

Version 0.16.2:
 - fixed install rule on case-insensitive filesystems

Version 0.16.1:
 - removed the custom definition of atan and used the one from Coq 8.4

Version 0.16.0:
 - switched build system to remake
 - moved to Coq 8.4

Version 0.15.1:
 - fixed failures when combining interval_intro with i_bisect*

Version 0.15.0:
 - added support for strict inequalities to the tactic
 - added support for integer power to the tactic
 - improved support for absolute value in the tactic
 - improved tactic so that it directly handles (e1 <= e2)

Version 0.14.0:
 - sped up square root for Z-based floating-point numbers
 - improved readability of the error messages for the tactic
 - modularized the tactic so that other specializations are available to user code
 - moved to Flocq 2.0

Version 0.13.0:
 - moved to Coq 8.3

Version 0.12.1:
 - fixed an incompatibility with Flocq 1.2

Version 0.12:
 - added a dependency on the Flocq library

Version 0.11:
 - removed i_nocheck parameter as computations are no longer done at Qed time


tactics.v
Require Import Reals.
Require Import List.
Require Import missing.
Require Import ZArith.
Require Import BigN.
Require Import BigZ.
Require Import xreal.
Require Import definitions.
Require Import generic.
Require Import generic_proof.
(*Require Import stdz_carrier.*)
Require Import bigint_carrier.
Require Import specific_ops.
Require Import interval.
Require Import interval_float_full.
Require Import bisect.

(*Module C := StdZRadix2.*)
Module C := BigIntRadix2.
Module F := SpecificFloat C.
Module I := FloatIntervalFull F.
Module V := Valuator I.

Ltac get_float t :=
  let get_mantissa t :=
    let rec aux t :=
      match t with
      | 1%R => xH
      | 2%R => constr:(xO xH)
      | 3%R => constr:(xI xH)
      | (2 * ?v)%R =>
        let w := aux v in constr:(xO w)
      | (1 + 2 * ?v)%R =>
        let w := aux v in constr:(xI w)
      end in
    aux t in
  let get_float_rational s n d :=
    let rec aux t :=
      match t with
      | 2%R => xH
      | (2 * ?v)%R =>
        let w := aux v in
        eval vm_compute in (Psucc w)
      end in
    let e := aux d in
    let m := get_mantissa n in
    eval vm_compute in (F.fromF (generic.Float 2 s m (Zneg e))) in
  let get_float_integer s t :=
    let rec aux m e :=
      match m with
      | xO ?v =>
        let u := eval vm_compute in (Zsucc e) in
        aux v u
      | _ => constr:(m, e)
      end in
    let m := get_mantissa t in
    let v := aux m Z0 in
    match v with
    | (?m, ?e) => eval vm_compute in (F.fromF (generic.Float 2 s m e))
    end in
  match t with
  | 0%R => F.zero
  | (-?n * /?d)%R => get_float_rational true n d
  | (?n * /?d)%R => get_float_rational false n d
  | (-?n / ?d)%R => get_float_rational true n d
  | (?n / ?d)%R => get_float_rational false n d
  | (-?n)%R => get_float_integer true n
  | ?n => get_float_integer false n
  | _ => false
  end.

Lemma Rabs_contains :
  forall f v,
  contains (I.convert (I.bnd (F.neg f) f)) (Xreal v) ->
  match F.toF f with
  | generic.Float false m e => (Rabs v <= FtoR F.radix false m e)%R
  | _ => True
  end.
intros [|m e] v (H1, H2).
exact I.
case_eq (F.toF (Float m e)) ; trivial.
intro b. case b ; trivial.
clear b.
intros.
apply Rabs_def1_le.
unfold I.I.convert_bound in H2.
rewrite H in H2.
exact H2.
unfold I.I.convert_bound in H1.
rewrite F.neg_correct in H1.
rewrite Fneg_correct in H1.
rewrite H in H1.
exact H1.
Qed.

Lemma Rabs_contains_rev :
  forall f v,
  match F.toF f with
  | generic.Float false m e => (Rabs v <= FtoR F.radix false m e)%R
  | _ => False
  end ->
  contains (I.convert (I.bnd (F.neg f) f)) (Xreal v).
intros [|m e] v.
do 2 split.
case_eq (F.toF (Float m e)) ; try (intros ; elim H0).
intros b p z.
case b ; intros.
elim H0.
generalize (Rabs_def2_le _ _ H0).
clear H0. intros (H1, H2).
split ; unfold I.I.convert_bound.
rewrite F.neg_correct.
rewrite Fneg_correct.
rewrite H.
exact H1.
rewrite H.
exact H2.
Qed.

Inductive expr :=
  | Econst : nat -> expr
  | Eunary : V.unary_op -> expr -> expr
  | Ebinary : V.binary_op -> expr -> expr -> expr.

Ltac list_add a l :=
  let rec aux a l n :=
    match l with
    | nil        => constr:(n, cons a l)
    | cons a _   => constr:(n, l)
    | cons ?x ?l =>
      match aux a l (S n) with
      | (?n, ?l) => constr:(n, cons x l)
      end
    end in
  aux a l O.

Ltac remove_constants t l :=
  let rec aux t l :=
    match get_float t with
    | false =>
      let aux_u o a :=
        match aux a l with
        | (?u, ?l) => constr:(Eunary o u, l)
        end in
      let aux_b o a b :=
        match aux b l with
        | (?v, ?l) =>
          match aux a l with
          | (?u, ?l) => constr:(Ebinary o u v, l)
          end
        end in
      match t with
      | Ropp ?a => aux_u V.Neg a
      | Rabs ?a => aux_u V.Abs a
      | Rinv ?a => aux_u V.Inv a
      | Rsqr ?a => aux_u V.Sqr a
      | Rmult ?a ?a => aux_u V.Sqr a
      | sqrt ?a => aux_u V.Sqrt a
      | cos ?a => aux_u V.Cos a
      | sin ?a => aux_u V.Sin a
      | tan ?a => aux_u V.Tan a
      | atan ?a => aux_u V.Atan a
      | exp ?a => aux_u V.Exp a
      | Rplus ?a ?b => aux_b V.Add a b
      | Rminus ?a ?b => aux_b V.Sub a b
      | Rplus ?a (Ropp ?b) => aux_b V.Sub a b
      | Rmult ?a ?b => aux_b V.Mul a b
      | Rdiv ?a ?b => aux_b V.Div a b
      | Rmult ?a (Rinv ?b) => aux_b V.Div a b
      | _ =>
        match list_add t l with
        | (?n, ?l) => constr:(Econst n, l)
        end
      end
    | _ =>
      match list_add t l with
      | (?n, ?l) => constr:(Econst n, l)
      end
    end in
  aux t l.

Ltac list_find1 a l :=
  let rec aux l n :=
    match l with
    | nil       => false
    | cons a _  => n
    | cons _ ?l => aux l (S n)
    end in
  aux l O.

Ltac get_non_constants t :=
  let rec aux t l :=
    match t with
    | Econst _ => l
    | _ =>
      match list_find1 t l with
      | false =>
        let m :=
          match t with
          | Eunary _ ?a => aux a l
          | Ebinary _ ?a ?b => aux a ltac:(aux b l)
          end in
        constr:(cons t m)
      | _ => l
      end
    end in
  aux t (@nil expr).

Ltac list_find2 a l :=
  let rec aux l n :=
    match l with
    | nil       => false
    | cons a _  => n
    | cons _ ?l => aux l (S n)
    end in
  match a with
  | Econst ?n => eval compute in (n + length l)%nat
  | _ => aux l O
  end.

Ltac generate_machine l :=
  let rec aux l q :=
    match l with
    | nil => q
    | cons ?t ?l =>
      let m :=
        match t with
        | Eunary ?o ?a =>
          let u := list_find2 a l in
          constr:(V.Unary o u)
        | Ebinary ?o ?a ?b =>
          let u := list_find2 a l in
          let v := list_find2 b l in
          constr:(V.Binary o u v)
        end in
      aux l (cons m q)
    end in
  aux l (@nil V.term).

Ltac extract_algorithm t l :=
  match remove_constants t l with
  | (?t, ?lc) =>
    let lm := generate_machine ltac:(get_non_constants t) in
    constr:(lm, lc)
  end.

Ltac xalgorithm lx :=
  match goal with
  | |- Rle ?a ?b /\ Rle ?b ?c =>
    let v := get_float a in
    let w := get_float c in
    change (contains (I.convert (I.bnd v w)) (Xreal b))
  | |- Rle (Rabs ?a) ?b =>
    let v := get_float b in
    match v with
    | Float ?m ?e => refine (Rabs_contains v a _)
    end
  | |- Rle ?a ?b =>
    let v := get_float b in
    refine (proj2 (_ : contains (I.convert (I.bnd F.nan v)) (Xreal a)))
  | |- Rle ?a ?b =>
    let v := get_float a in
    refine (proj1 (_ : contains (I.convert (I.bnd v F.nan)) (Xreal b)))
  | _ => fail 100 "Goal is not an inequality with floating-point bounds."
  end ;
  match goal with
  | |- contains ?xi (Xreal ?y) =>
    match extract_algorithm y lx with
    | (?a, ?b) =>
      let formula := fresh "formula" in
      pose (formula := a) ;
      refine (V.xreal_to_real formula b O xi _)
    end
  end.

Ltac get_bounds l :=
  let rec aux l :=
    match l with
    | nil => constr:(@nil V.bound_proof)
    | cons ?x ?l =>
      let i :=
        match goal with
        | H: Rle ?a x /\ Rle x ?b |- _ =>
          let v := get_float a in
          let w := get_float b in
          constr:(V.Bproof x (I.bnd v w) H)
        | H: Rle ?a x |- _ =>
          let v := get_float a in
          constr:(V.Bproof x (I.bnd v F.nan) (conj H I))
        | H: Rle x ?b |- _ =>
          let v := get_float b in
          constr:(V.Bproof x (I.bnd F.nan v) (conj I H))
        | H: Rle (Rabs x) ?b |- _ =>
          let v := get_float b in
          match v with
          | Float ?m ?e =>
            constr:(V.Bproof x (I.bnd (F.neg v) v) (Rabs_contains_rev v x H))
          end
        | _ =>
          let v := get_float x in
          constr:(let f := v in V.Bproof x (I.bnd f f) (conj (Rle_refl x) (Rle_refl x)))
        | _ => fail 100 "Atom" x "is neither a floating-point value nor bounded by floating-point values."
        end in
      let m := aux l in
      constr:(cons i m)
    end in
  aux l.

Lemma interval_helper_evaluate :
  forall bounds output formula prec n,
  I.subset (nth n (V.eval_bnd prec formula (map V.interval_from_bp bounds)) I.nai) output = true ->
  contains (I.convert output) (nth n (V.eval_ext formula (map V.xreal_from_bp bounds)) (Xreal 0)).
intros.
eapply subset_contains.
apply I.subset_correct.
apply H.
apply V.eval_bnd_correct.
Qed.

Lemma interval_helper_bisection :
  forall bounds output formula prec depth n,
  match bounds with
  | cons (V.Bproof _ (interval_float.Ibnd l u) _) tail =>
    V.Algos.bisect_1d (fun b => nth n (V.eval_bnd prec formula (b :: map V.interval_from_bp tail)) I.nai) l u output depth = true
  | _ => False
  end ->
  contains (I.convert output) (nth n (V.eval_ext formula (map V.xreal_from_bp bounds)) (Xreal 0)).
intro.
case bounds.
intros.
elim H.
intro.
case b.
intros x xi.
case xi.
intros.
elim H.
intros.
clear bounds b xi.
pose (f := fun x => nth n (V.eval_ext formula (x :: map V.xreal_from_bp l0)) (Xreal 0)).
pose (fi := fun b => nth n (V.eval_bnd prec formula (b :: map V.interval_from_bp l0)) I.nai).
change (contains (I.convert output) (f (Xreal x))).
refine (V.Algos.bisect_1d_correct depth f fi _ _ _ _ H _ c).
exact (V.eval_bnd_correct_ext _ _ _ _).
Qed.

Lemma interval_helper_bisection_diff :
  forall bounds output formula prec depth n,
  match bounds with
  | cons (V.Bproof _ (interval_float.Ibnd l u) _) tail =>
    V.Algos.bisect_1d (fun b => V.eval_diff prec formula (map V.interval_from_bp tail) n b) l u output depth = true
  | _ => False
  end ->
  contains (I.convert output) (nth n (V.eval_ext formula (map V.xreal_from_bp bounds)) (Xreal 0)).
intro.
case bounds.
intros.
elim H.
intro.
case b.
intros x xi.
case xi.
intros.
elim H.
intros.
clear bounds b xi.
pose (f := fun x => nth n (V.eval_ext formula (x :: map V.xreal_from_bp l0)) (Xreal 0)).
pose (fi := fun b => V.eval_diff prec formula (map V.interval_from_bp l0) n b).
change (contains (I.convert output) (f (Xreal x))).
refine (V.Algos.bisect_1d_correct depth f fi _ _ _ _ H _ c).
exact (V.eval_diff_correct_ext _ _ _ _).
Qed.

(*
Lemma interval_helper_bisection_first_order :
  forall bounds output formula prec depth n,
  match bounds with
  | cons (V.Bproof _ (interval_float.Ibnd l u) _) tail =>
    V.Algos.bisect_1d (fun b => V.eval_first_order_nth prec formula (map V.interval_from_bp tail) n b) l u output depth = true
  | _ => False
  end ->
  contains (I.convert output) (nth n (V.eval_ext formula (map V.xreal_from_bp bounds)) (Xreal 0)).
intro.
case bounds.
intros.
elim H.
intro.
case b.
intros x xi.
case xi.
intros.
elim H.
intros.
clear bounds b xi.
pose (f := fun x => nth n (V.eval_ext formula (x :: map V.xreal_from_bp l0)) (Xreal 0)).
pose (fi := fun b => V.eval_first_order_nth prec formula (map V.interval_from_bp l0) n b).
change (contains (I.convert output) (f (Xreal x))).
refine (V.Algos.bisect_1d_correct depth f fi _ _ _ _ H _ c).
exact (V.eval_first_order_correct_ext _ _ _ _).
Qed.
*)

Ltac do_interval vars prec depth eval_tac check_tac :=
  match goal with
  | |- contains (I.convert _) (nth _ (V.eval_ext _ (map Xreal _)) (Xreal 0)) => idtac
  | _ => xalgorithm vars
  end ;
  match goal with
  | |- contains (I.convert ?output) (nth ?n
      (V.eval_ext ?formula (map Xreal ?constants)) (Xreal 0)) =>
    let bounds_ := get_bounds constants in
    let bounds := fresh "bounds" in
    pose (bounds := bounds_) ;
    let prec := eval vm_compute in (C.ZtoE (Z_of_nat prec)) in
    change (map Xreal constants) with (map V.xreal_from_bp bounds) ;
    eval_tac bounds output formula prec depth n ;
    check_tac
  end.

Ltac do_interval_check :=
  vm_compute ;
  try exact (refl_equal true) ;
  fail 100 "Numerical evaluation failed to conclude. You may want to adjust some parameters.".

Ltac do_interval_nocheck :=
  vm_cast_no_check (refl_equal true).

Ltac do_interval_eval bounds output formula prec depth n :=
  refine (interval_helper_evaluate bounds output formula prec n _).

Ltac do_interval_bisect bounds output formula prec depth n :=
  refine (interval_helper_bisection bounds output formula prec depth n _).

Ltac do_interval_bisect_diff bounds output formula prec depth n :=
  refine (interval_helper_bisection_diff bounds output formula prec depth n _).

Inductive interval_tac_parameters :=
  | i_prec : nat -> interval_tac_parameters
  | i_nocheck : interval_tac_parameters
  | i_bisect : R -> interval_tac_parameters
  | i_bisect_diff : R -> interval_tac_parameters
  | i_depth : nat -> interval_tac_parameters.

Ltac tuple_to_list params l :=
  match params with
  | pair ?a ?b => tuple_to_list a (b :: l)
  | ?b => constr:(b :: l)
  end.

Ltac do_interval_parse params :=
  let rec aux vars prec depth eval_tac check_tac params :=
    match params with
    | nil => do_interval vars prec depth eval_tac check_tac
    | cons (i_prec ?p) ?t => aux vars p depth eval_tac check_tac t
    | cons i_nocheck ?t => aux vars prec depth eval_tac do_interval_nocheck t
    | cons (i_bisect ?x) ?t => aux (cons x nil) prec depth do_interval_bisect check_tac t
    | cons (i_bisect_diff ?x) ?t => aux (cons x nil) prec depth do_interval_bisect_diff check_tac t
    | cons (i_depth ?d) ?t => aux vars prec d eval_tac check_tac t
    | cons ?h _ => fail 100 "Unknown tactic parameter" h "."
    end in
  aux (@nil R) 30%nat 15%nat do_interval_eval do_interval_check params.

Tactic Notation "interval" :=
  do_interval_parse (@nil interval_tac_parameters).

Tactic Notation "interval" "with" constr(params) :=
  do_interval_parse ltac:(tuple_to_list params (@nil interval_tac_parameters)).

Ltac do_interval_generalize t b :=
  match eval vm_compute in (I.convert b) with
  | Inan => fail 100 "Nothing known about" t
  | Ibnd ?l ?u =>
    match goal with
    | |- ?P =>
      match l with
      | Xnan =>
        match u with
        | Xnan => fail 100 "Nothing known about" t
        | Xreal ?u => refine ((_ : (t <= u)%R -> P) _)
        end
      | Xreal ?l =>
        match u with
        | Xnan => refine ((_ : (l <= t)%R -> P) _)
        | Xreal ?u => refine ((_ : (l <= t <= u)%R -> P) _)
        end
      end
    end
  end.

Ltac do_interval_intro_eval extend bounds formula prec depth :=
  eval vm_compute in (extend (nth 0 (V.eval_bnd prec formula (map V.interval_from_bp bounds)) I.nai)).

Ltac do_interval_intro_bisect extend bounds formula prec depth :=
  eval vm_compute in
   (match bounds with
    | cons (V.Bproof _ (interval_float.Ibnd l u) _) tail =>
      V.Algos.lookup_1d (fun b => nth 0 (V.eval_bnd prec formula (b :: map V.interval_from_bp tail)) I.nai) l u extend depth
    | _ => I.nai
    end).

Ltac do_interval_intro_bisect_diff extend bounds formula prec depth :=
  eval vm_compute in
   (match bounds with
    | cons (V.Bproof _ (interval_float.Ibnd l u) _) tail =>
      V.Algos.lookup_1d (fun b => V.eval_diff prec formula (map V.interval_from_bp tail) 0 b) l u extend depth
    | _ => I.nai
    end).

Ltac do_interval_intro t extend params vars prec depth eval_tac :=
  let prec := eval vm_compute in (C.ZtoE (Z_of_nat prec)) in
  match extract_algorithm t vars with
  | (?formula, ?constants) =>
    let bounds := get_bounds constants in
    let v := eval_tac extend bounds formula prec depth in
    do_interval_generalize t v ;
    [ intro | do_interval_parse params ]
  end.

Ltac do_interval_intro_parse t_ extend params_ :=
  let rec aux vars prec depth eval_tac params :=
    match params with
    | nil => do_interval_intro t_ extend params_ vars prec depth eval_tac
    | cons (i_prec ?p) ?t => aux vars p depth eval_tac t
    | cons i_nocheck ?t => aux vars prec depth eval_tac t
    | cons (i_bisect ?x) ?t => aux (cons x nil) prec depth do_interval_intro_bisect t
    | cons (i_bisect_diff ?x) ?t => aux (cons x nil) prec depth do_interval_intro_bisect_diff t
    | cons (i_depth ?d) ?t => aux vars prec d eval_tac t
    | cons ?h _ => fail 100 "Unknown tactic parameter" h "."
    end in
  aux (@nil R) 30%nat 5%nat do_interval_intro_eval params_.

Tactic Notation "interval_intro" constr(t) :=
  do_interval_intro_parse t (fun v : I.type => v) (cons i_nocheck nil).

Tactic Notation "interval_intro" constr(t) "lower" :=
  do_interval_intro_parse t I.upper_extent (cons i_nocheck nil).

Tactic Notation "interval_intro" constr(t) "upper"  :=
  do_interval_intro_parse t I.lower_extent (cons i_nocheck nil).

Tactic Notation "interval_intro" constr(t) "with" constr(params) :=
  do_interval_intro_parse t (fun v : I.type => v) ltac:(tuple_to_list params (cons i_nocheck nil)).

Tactic Notation "interval_intro" constr(t) "lower" "with" constr(params) :=
  do_interval_intro_parse t I.upper_extent ltac:(tuple_to_list params (cons i_nocheck nil)).

Tactic Notation "interval_intro" constr(t) "upper" "with" constr(params) :=
  do_interval_intro_parse t I.lower_extent ltac:(tuple_to_list params (cons i_nocheck nil)).

(*
Lemma blo1 :
  forall x, (Rabs x <= 5)%R ->
  (-4 <= x + 1)%R.
intros.
interval.
Qed.
*)

(*
Lemma blo2 :
  (2/3 <= 5/7)%R.
intros.
interval_intro (5/7)%R lower with (i_prec 4%nat).
apply Rle_trans with (2 := H).
interval.
Qed.

Print blo2.
*)

(*
Lemma blo3 :
  forall x, (x <= 0)%R ->
  (0 <= x - x <= 0)%R.
intros.
Time interval with (i_bisect_diff x).
Qed.
*)

(*
Lemma blo4 :
  forall x, (3/2 <= x <= 2)%R ->
  forall y, (1 <= y <= 33/32)%R ->
  (Rabs (sqrt(1 + x/sqrt(x+y)) - 144/1000*x - 118/100) <= 71/32768)%R.
intros.
interval with (i_bisect x).
Qed.
*)
